core/consensus/common/primitives/src/lib.rs

```diff
@@ -29,3 +31,45 @@ decl_runtime_apis! {
 		fn authorities() -> Vec<AuthorityId>;
 	}
 }
+
+pub trait AuthorshipEquivocationProof {
```

Comment:
IMO this is not needed, it's not abstracting over anything anyway.

Reply:
It's abstracting over Aura and Babe equivocation proof (I have the Aura implementation with low priority).

Reply:
https://github.com/paritytech/substrate/pull/3314/files/1a88f9016ae3b16937afce5e3fb189749dd8e521#diff-1144b2e008f250f4ca22726b35ee22c7R65

How will you implement this for GRANDPA? Same goes for some other methods which are specific to block proposal equivocations. My point is there's too little in common between BABE and GRANDPA equivocations to abstract anything, so there's no point in abstracting. They'll be reported to different modules anyway so I don't know why you need a common type.

Reply:
It's interfacing  Aura and Babe with the Slots module. If I remove this I need to duplicate code in Slots module, or there is a better way?

Reply:
Sorry I misread your comment earlier and thought this was supposed to abstract over GRANDPA and BABE.